kube-core/src/duration.rs

```diff
@@ -0,0 +1,430 @@
+
+        let mut total = time::Duration::from_secs(0);
+        while !s.is_empty() {
+            if let Some(unit_start) = s.find(|c: char| c.is_alphabetic()) {
+                let (val, rest) = s.split_at(unit_start);
+                let val = val.parse::<f64>()?;
+                let unit = if let Some(next_numeric_start) = rest.find(|c: char| !c.is_alphabetic()) {
+                    let (unit, rest) = rest.split_at(next_numeric_start);
+                    s = rest;
+                    unit
+                } else {
+                    s = "";
+                    rest
+                };
+                total += duration_from_units(val, unit)?;
+            } else if s == "0" {
+                return Ok(Duration { duration: time::Duration::from_secs(0), is_negative });
+            } else {
+                return Err(ParseError::NoUnit);
+            }
+        }
+
+        Ok(Duration { duration: total, is_negative })
+
+    }
+}
+
+impl Serialize for Duration {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_str(self)
+    }
+}
+
+impl<'de> Deserialize<'de> for Duration {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>
+    {
+        struct Visitor;
+        impl<'de> de::Visitor<'de> for Visitor {
+            type Value = Duration;
+
+            fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                f.write_str("a string in Go `time.time::Duration.String()` format")
+            }
+
+
+            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                let val = value.parse::<Duration>().map_err(de::Error::custom)?;
+                Ok(val)
+            }
+        }
+        deserializer.deserialize_str(Visitor)
+    }
+}
+
+impl PartialEq<time::Duration> for Duration {
+    fn eq(&self, other: &time::Duration) -> bool {
+        // Since `std::time::Duration` is unsigned, a negative `Duration` is
+        // never equal to a `std::time::Duration`.
+        if self.is_negative {
+            return false;
+        }
+
+        self.duration == *other
+    }
+}
+
+impl PartialEq<time::Duration> for &'_ Duration {
+    fn eq(&self, other: &time::Duration) -> bool {
+        // Since `std::time::Duration` is unsigned, a negative `Duration` is
+        // never equal to a `std::time::Duration`.
+        if self.is_negative {
+            return false;
+        }
+
+        self.duration == *other
+    }
+}
+
+impl PartialEq<Duration> for time::Duration {
+    fn eq(&self, other: &Duration) -> bool {
+        // Since `std::time::Duration` is unsigned, a negative `Duration` is
+        // never equal to a `std::time::Duration`.
+        if other.is_negative {
+            return false;
+        }
+
+        self == other.duration
```

Comment:
do we need a ref here?

```suggestion
        self == &other.duration
```